File: R/write.R


setMethod("blocks", signature(x="SpatRaster"),
	function(x, n=4) {
		opt <- spatOptions("", FALSE, ncopies=n)
		b <- x@ptr$getBlockSizeR(n, opt$memfrac)
		b$row <- b$row + 1
		b
	}
)


setMethod("writeStart", signature(x="SpatRaster", filename="character"),
	function(x, filename="", overwrite=FALSE, n=4, sources="", ...) {
		filename <- path.expand(trimws(filename[1]))
		filename <- enc2utf8(filename)
		opt <- spatOptions(filename, overwrite, ncopies=n, ...)
		ok <- x@ptr$writeStart(opt, unique(sources))
		messages(x, "writeStart")
		b <- x@ptr$getBlockSizeWrite()
		b$row <- b$row + 1
		b
	}
)


setMethod("writeStop", signature(x="SpatRaster"),
	function(x) {
		success <- x@ptr$writeStop()
		messages(x, "writeStop")
		f <- sources(x)
		if (f != "") {
			x <- rast(f)
		}
		return(x)
	}
)

setMethod("writeValues", signature(x="SpatRaster", v="vector"),
	function(x, v, start, nrows) {
		success <- x@ptr$writeValues(v, start-1, nrows)
		messages(x, "writeValues")
		invisible(success)
	}
)


setMethod("writeRaster", signature(x="SpatRaster", filename="character"),
function(x, filename="", overwrite=FALSE, ...) {
	filename <- trimws(filename)
	stopifnot(filename != "")
	filename <- enc2utf8(filename)

	ftp <- list(...)$filetype[1]
	fext <- tools::file_ext(filename)
	if (any(fext %in% c("nc", "cdf")) && (is.null(ftp) || isTRUE(ftp=="netCDF"))) {
		warn("writeRaster", "consider writeCDF to write ncdf files")
	} 
	#else if ((length(fext)==1) && (any(fext == "rds")) && ((is.null(ftp)) || isTRUE(ftp=="RDS"))) {
	#	if ((!overwrite) && file.exists(filename)) {
	#		error("writeRaster", "Use 'overwrite=TRUE' to overwrite an existing file")
	#	}
	#	return(saveRDS(x, filename))
	#}
	opt <- spatOptions(filename, overwrite, ...)
	x@ptr <- x@ptr$writeRaster(opt)
	x <- messages(x, "writeRaster")
	invisible(rast(filename))
}
)


get_filetype <- function(filename) {  
	fn <- tolower(filename)
	ext <- tools::file_ext(fn)
	ext2 <- gsub("^[^.]*\\.(.*).*$", "\\1", fn)
	if (ext == "shp" || ext == "shz" || (ext == "zip" && ext2 == "shp.zip") || ext == "") {
		"ESRI Shapefile"
	} else if (ext == "gpkg" || (ext == "zip" && ext2 == "gpkg.zip")) {
		"GPKG"
	} else if (ext == "gdb") {
		"OpenFileGDB"
	} else if (ext == "gml") {
		"GML"
	} else if (ext == "json" || ext == "geojson") {
		"GeoJSON"
	} else if (ext == "cdf") {
		"netCDF"
	} else if (ext == "svg") {
		"SVG"
	} else if (ext == "kml") {
		"KML"
	} else if (ext == "vct") {
		"Idrisi"
	} else if (ext == "rds") {
		"rds"
	} else if (ext == "tab") {
		"MapInfo File"
	} else {
		error("writeVector", "cannot guess filetype from filename")
	}
}

setMethod("writeVector", signature(x="SpatVector", filename="character"),
function(x, filename, filetype=NULL, layer=NULL, insert=FALSE, overwrite=FALSE, options="ENCODING=UTF-8") {

	filename <- path.expand(trimws(filename[1]))
	filename <- enc2utf8(filename)
	if (filename == "") {
		error("writeVector", "provide a filename")
	}
	if (is.null(filetype)) {
		filetype <- get_filetype(filename)
		if (filetype == "rds") {
			return(saveRDS(x, filename))
		}
	}
	if (is.null(layer)) layer <- tools::file_path_sans_ext(basename(filename))
	layer <- trimws(layer)

	if (is.null(options)) { options <- ""[0] }

	if (filetype == "ESRI Shapefile") {
		nms <- names(x)
		i <- nchar(nms) > 10
		if (any(i)) {
			nms[i] <- substr(nms[i], 1, 10)
			testnms <- make.unique(nms, sep="")
			if (!all(testnms == nms)) {
				i <- which(i)
				newnms <- substr(nms[i], 1, 9)
				newnms <- make.unique(newnms, sep="")
				j <- which(nchar(newnms) == 9)
				newnms[j] <- paste0(newnms[j], "0")
				nms[i] <- newnms
			}
			x@ptr <- x@ptr$deepcopy()
			names(x) <- nms
		}
	}
	success <- x@ptr$write(filename, layer, filetype, insert[1], overwrite[1], options)
	messages(x, "writeVector")
	invisible(TRUE)
}
)



# setMethod("writeVector", signature(x="SpatVectorProxy", filename="character"),
# function(x, filename, filetype=NULL, layer=NULL, insert=FALSE, overwrite=FALSE, options="ENCODING=UTF-8") {
	# filename <- trimws(filename)
	# filename <- enc2utf8(filename)
	# if (filename == "") {
		# error("writeVector", "provide a filename")
	# }
	# if (is.null(filetype)) {
		# filetype <- get_filetype(filename)
	# }
	# if (is.null(layer)) layer <- tools::file_path_sans_ext(basename(filename))
	# layer <- trimws(layer)

	# if (is.null(options)) { options <- ""[0] }

	# if (filetype == "ESRI Shapefile") {
		# nms <- names(x)
		# i <- nchar(nms) > 10
		# if (any(i)) {
			# nms[i] <- substr(nms[i], 1, 10)
			# testnms <- make.unique(nms, sep="")
			# if (!all(testnms == nms)) {
				# i <- which(i)
				# newnms <- substr(nms[i], 1, 9)
				# newnms <- make.unique(newnms, sep="")
				# j <- which(nchar(newnms) == 9)
				# newnms[j] <- paste0(newnms[j], "0")
				# nms[i] <- newnms
			# }
			# x@ptr <- x@ptr$deepcopy()
			# names(x) <- nms
		# }
	# }
	# success <- x@ptr$write_proxy(filename, layer, filetype, insert[1], FALSE, overwrite[1], options)
	# messages(x, "writeVector")
	# invisible(TRUE)
# }
# )

File: python/src/terra/write.py
"""
write.py — write raster and vector data to files.
"""
from __future__ import annotations
import os
from typing import List, Optional, Union

from ._terra import SpatRaster, SpatVector, SpatOptions
from ._helpers import messages, spatoptions

_cpp_vect_write = SpatVector.write  # captured before monkey-patching


def _opt() -> SpatOptions:
    return SpatOptions()


# ---------------------------------------------------------------------------
# SpatRaster write
# ---------------------------------------------------------------------------

def write_raster(
    x: SpatRaster,
    filename: str,
    overwrite: bool = False,
    filetype: Optional[str] = None,
    datatype: str = "FLT4S",
    gdal: Optional[List[str]] = None,
    **kwargs,
) -> SpatRaster:
    """
    Write a SpatRaster to a file.

    Parameters
    ----------
    x : SpatRaster
    filename : str
        Output path.  The format is inferred from the extension
        (e.g. ``".tif"`` → GeoTIFF, ``".nc"`` → NetCDF).
    overwrite : bool
        If False (default), raise an error if *filename* already exists.
    filetype : str, optional
        GDAL driver name (e.g. ``"GTiff"``).  Auto-detected if None.
    datatype : str
        Output data type: ``"FLT4S"`` (float32, default), ``"FLT8S"``
        (float64), ``"INT2S"`` (int16), ``"INT4S"`` (int32), ``"INT1U"``
        (uint8), etc.
    gdal : list of str, optional
        GDAL creation options (e.g. ``["COMPRESS=LZW"]``).
    **kwargs
        Additional options forwarded to SpatOptions.

    Returns
    -------
    SpatRaster  (re-opened from the written file)
    """
    filename = os.path.expanduser(filename.strip())
    if not filename:
        raise ValueError("provide a filename")
    opt = spatoptions(filename, overwrite)
    opt.datatype = datatype
    if filetype:
        opt.filetype = filetype
    if gdal:
        opt.gdal_options = gdal
    xc = x.writeRaster(opt)
    messages(xc, "writeRaster")
    from .rast import rast
    return rast(filename)


def write_start(
    x: SpatRaster,
    filename: str,
    overwrite: bool = False,
    n: int = 4,
    sources: Optional[List[str]] = None,
) -> dict:
    """
    Open a file for block-wise writing.

    Parameters
    ----------
    x : SpatRaster
    filename : str
    overwrite : bool
    n : int
        Number of block copies to buffer in memory.
    sources : list of str, optional

    Returns
    -------
    dict with keys ``"n"``, ``"row"``, ``"nrows"``.
    """
    filename = os.path.expanduser(filename.strip())
    opt = spatoptions(filename, overwrite)
    opt.ncopies = n
    if sources is None:
        sources = []
    ok = x.writeStart(opt, list(set(sources)))
    messages(x, "writeStart")
    b = x.getBlockSizeWrite()
    b["row"] = [r + 1 for r in b["row"]]
    return b


def write_values(
    x: SpatRaster,
    v: List[float],
    start: int,
    nrows: int,
) -> bool:
    """
    Write a block of values to a file opened with write_start().

    Parameters
    ----------
    x : SpatRaster
    v : list of float
    start : int
        Starting row (1-based).
    nrows : int
        Number of rows in this block.

    Returns
    -------
    bool
    """
    ok = x.writeValues(list(v), start - 1, nrows)
    messages(x, "writeValues")
    return bool(ok)


def write_stop(x: SpatRaster) -> SpatRaster:
    """
    Finalise a block-wise write and close the file.

    Parameters
    ----------
    x : SpatRaster

    Returns
    -------
    SpatRaster  (re-opened from the written file)
    """
    x.writeStop()
    messages(x, "writeStop")
    src = list(x.filenames())
    if src and src[0]:
        from .rast import rast
        return rast(src[0])
    return x


def blocks(x: SpatRaster, n: int = 4) -> dict:
    """
    Return the block structure for writing *x*.

    Parameters
    ----------
    x : SpatRaster
    n : int
        Number of copies to buffer.

    Returns
    -------
    dict with keys ``"n"``, ``"row"`` (1-based), ``"nrows"``.
    """
    opt = spatoptions()
    opt.ncopies = n
    b = x.getBlockSizeR(opt)
    b["row"] = [r + 1 for r in b["row"]]
    return b


# ---------------------------------------------------------------------------
# SpatVector write
# ---------------------------------------------------------------------------

_EXT_TO_FILETYPE = {
    "shp": "ESRI Shapefile",
    "shz": "ESRI Shapefile",
    "gpkg": "GPKG",
    "gdb": "OpenFileGDB",
    "gml": "GML",
    "json": "GeoJSON",
    "geojson": "GeoJSON",
    "cdf": "netCDF",
    "svg": "SVG",
    "kml": "KML",
    "vct": "Idrisi",
    "tab": "MapInfo File",
}


def _guess_filetype(filename: str) -> str:
    ext = os.path.splitext(filename)[1].lstrip(".").lower()
    ft = _EXT_TO_FILETYPE.get(ext)
    if ft is None:
        raise ValueError(f"cannot guess filetype from filename {filename!r}")
    return ft


def write_vector(
    x: SpatVector,
    filename: str,
    filetype: Optional[str] = None,
    layer: Optional[str] = None,
    insert: bool = False,
    overwrite: bool = False,
    options: str = "ENCODING=UTF-8",
) -> bool:
    """
    Write a SpatVector to a file.

    Parameters
    ----------
    x : SpatVector
    filename : str
        Output path.
    filetype : str, optional
        OGR driver name.  Auto-detected from extension if None.
    layer : str, optional
        Layer name inside the file.  Defaults to the base filename.
    insert : bool
        Append to an existing file instead of replacing.
    overwrite : bool
        Overwrite an existing layer.
    options : str
        OGR creation options string.

    Returns
    -------
    bool
    """
    filename = os.path.expanduser(filename.strip())
    if not filename:
        raise ValueError("provide a filename")
    if filetype is None:
        filetype = _guess_filetype(filename)
    if layer is None:
        layer = os.path.splitext(os.path.basename(filename))[0].strip()

    # Truncate field names for Shapefiles (max 10 chars)
    if filetype == "ESRI Shapefile":
        nms = list(x.names)
        truncated = [n[:10] for n in nms]
        if truncated != nms:
            xc = x.deepcopy()
            xc.names = truncated
            x = xc

    ok = _cpp_vect_write(x, filename, layer, filetype, insert, overwrite, options)
    messages(x, "writeVector")
    return bool(ok)
